### DISCUSSION TITLE
Problem with "vscode Getting Started" guide: quarto preview is not selecting correct interpreter, incorrectly claims python env is unactivated

### DISCUSSION THREAD

**Discussion by angelotrivelli** [quarto-dev/quarto-cli#5140] [Q&A] [answered]
I am trying to follow the "getting started" guide for quarto with vscode and cannot get the first example to work. This one: [https://quarto.org/docs/get-started/hello/vscode.html](https://quarto.org/docs/get-started/hello/vscode.html).

I am on a windows 11 machine and manage my python interpreters with pyenv-win. The global default python is 3.11.2. 
  ```
PS C:\Users\angelo> python --version
  Python 3.11.2

PS C:\Users\angelo> pyenv versions
  3.10.9
  3.11.1
* 3.11.2 (set by C:\Users\angelo\.pyenv\pyenv-win\version)
```

I have installed quarto-cli and I've captured the output from `quarto check` (see below).
I don't have jupyter available globally. I assume that's OK and that I can just add it to my virtual environment. No need to install it system-wide, right?

```
PS C:\Users\angelo> quarto check

[>] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.1: OK
      Dart Sass version 1.55.0: OK
[>] Checking versions of quarto dependencies......OK
[>] Checking Quarto installation......OK
      Version: 1.3.313
      Path: C:\Users\angelo\AppData\Local\Programs\Quarto\bin
      CodePage: 1252

[>] Checking basic markdown render....OK

[>] Checking Python 3 installation....OK
      Version: 3.11.2
      Path: C:/Users/angelo/.pyenv/pyenv-win/versions/3.11.2/python3.exe
      Jupyter: (None)

      Jupyter is not available in this Python installation.
      Install with python -m pip install jupyter

[>] Checking R installation...........OK
      Version: 4.2.3
      Path: C:/PROGRA~1/R/R-42~1.3
      LibPaths:
        - C:/Users/angelo/AppData/Local/R/win-library/4.2
        - C:/Program Files/R/R-4.2.3/library
      knitr: 1.42
      rmarkdown: 2.20

[>] Checking Knitr engine render......OK
```

I also have vscode. It's already setup for python programming and today I installed the latest quarto extension, `Quarto v1.79.0`  and then created a project folder, 'quarto_python_vscode', launched vscode, and copy-pasted the `hello.qmd` file from [getting-started](https://quarto.org/docs/get-started/hello/vscode.html#render-and-preview). 

I've been using poetry to work with python projects, but I couldn't get it to work with this quarto example. So in order to increase the likelyhood of getting help, I am following the example in getting-started as closely as possible (and temporarily abandoning poetry). 

In the vscode terminal, I setup my local python with `pyenv local 3.11.2` and then the virtual environment, `python -m venv .venv` and then activate it (took a screenshot so it's clear that vscode has activated the virtual environment):

<img width="347" alt="image" src="https://user-images.githubusercontent.com/588784/230794486-72c0e414-14ba-42ca-b862-39f2678cbb3c.png">

Just as a sanity check, running `quarto check` yields this...
```
(.venv) PS C:\temp\quarto_python_vscode> quarto check

[>] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.1: OK
      Dart Sass version 1.55.0: OK
[>] Checking versions of quarto dependencies......OK
[>] Checking Quarto installation......OK
      Version: 1.3.313
      Path: C:\Users\angelo\AppData\Local\Programs\Quarto\bin
      CodePage: 1252

[>] Checking basic markdown render....OK

[>] Checking Python 3 installation....OK
      Version: 3.11.2
      Path: C:/Users/angelo/.pyenv/pyenv-win/versions/3.11.2/python3.exe
      Jupyter: (None)

      Jupyter is not available in this Python installation.
      Install with python -m pip install jupyter

      There is an unactivated Python environment in .venv. Did you forget to activate it?

[>] Checking R installation...........OK
      Version: 4.2.3
      Path: C:/PROGRA~1/R/R-42~1.3
      LibPaths:
        - C:/Users/angelo/AppData/Local/R/win-library/4.2
        - C:/Program Files/R/R-4.2.3/library
      knitr: 1.42
      rmarkdown: 2.20

[>] Checking Knitr engine render......OK
```
Well, that's not right! It says that my Python environment is unactivated **but I literally just activated it!** Moreover, it doesn't seem to be pointing to my .venv python. It's just pointing to the pyenv global python.

Going on to install dependencies anyway (in the activated virtual environment that quarto says is unactivated):
```
(.venv) PS C:\temp\quarto_python_vscode> python -m pip install jupyter matplotlib plotly
```
It seems to install, there's a boatload of stuff in .venv/Lib/. 

But when I try to render, it fails like below, again complaining about unactivated virtual environment, and also saying that jupyter is not available (_even though I just added it_).

```
PS C:\temp\quarto_python_vscode> & c:/temp/quarto_python_vscode/.venv/Scripts/Activate.ps1
(.venv) PS C:\temp\quarto_python_vscode> quarto preview c:/temp/quarto_python_vscode/hello.qmd --no-browser --no-watch-inputs

Starting python3 kernel...Traceback (most recent call last):
  File "C:\Users\angelo\AppData\Local\Programs\Quarto\share\jupyter\jupyter.py", line 21, in <module>
    from notebook import notebook_execute, RestartKernel
  File "C:\Users\angelo\AppData\Local\Programs\Quarto\share\jupyter\notebook.py", line 17, in <module>
    import nbformat
ModuleNotFoundError: No module named 'nbformat'
Python 3 installation:
  Version: 3.11.2
  Path: C:/Users/angelo/.pyenv/pyenv-win/versions/3.11.2/python3.exe
  Jupyter: (None)

Jupyter is not available in this Python installation.
Install with python -m pip install jupyter

There is an unactivated Python environment in .venv. Did you forget to activate it?

(.venv) PS C:\temp\quarto_python_vscode> 
```

It looks like quarto is not able to detect my python virtual environment?

I would really like to use quarto. It seems mostly OK for me in rstudio, and it would be so awesome to use it in vscode + python as well. But I REALLY NEED to use a virtual environment with python stuff. Of course, it would be necessary to get this "out-of-the-box" hello-world example working as well.

Does anyone have ideas about what might be going on here and how I can troubleshoot this further?

**Comment by bassamsdata** [comment]
well, I have the same issue here on mac with conda, I tried multiple solutions even the global variable `QUARTO_PYTHON`

quarto-cli and quarto in VSCode insist on using `python 3.8` on my mac, eventhough I have specified another version 
`export QUARTO_PYTHON=/usr/local/opt/python@3.11/bin/python3.11`

I couldn't able to solve the issue.
I have the latest version of everything (quarto 1.3.340).

I hope someone from the quarto team look into this issue. I'm not sure if @angelotrivelli found a solution for that.

**Reply by cscheid** [reply]
> quarto-cli and quarto in VSCode insist on using python 3.8 on my mac

What's probably happening is that you're specifying `export QUARTO_PYTHON=/usr/local/opt/python@3.11/bin/python3.11` in a file that is not processed by your shell. You can test this by rendering the following file:

````
---
title: env test
---

```{python}
import os
print(os.getenv("QUARTO_PYTHON"))
```
````

And then check to see that the output is what you expect.

**Comment by angelotrivelli** [comment]
Unfortunately, I still get the same problem even after updating to Quarto `1.3.340`.

To summarize, even after activating my virtual environment with 
```
    .\.venv\Scripts\Activate.ps1
```

Quarto preview fails to use the venv python installation (and the libraries I had installed into it like jupyter):

```
(.venv) PS C:\projects\quarto_python_vscode> quarto preview c:/projects/quarto_python_vscode/hello.qmd --no-browser --no-watch-inputs

Starting python3 kernel...Traceback (most recent call last):
  File "C:\Users\angelo\AppData\Local\Programs\Quarto\share\jupyter\jupyter.py", line 21, in <module>
    from notebook import notebook_execute, RestartKernel
  File "C:\Users\angelo\AppData\Local\Programs\Quarto\share\jupyter\notebook.py", line 17, in <module>
    import nbformat
ModuleNotFoundError: No module named 'nbformat'
Python 3 installation:
  Version: 3.11.2
  Path: C:/Users/angelo/.pyenv/pyenv-win/versions/3.11.2/python3.exe
  Jupyter: (None)

Jupyter is not available in this Python installation.
Install with python -m pip install jupyter

There is an unactivated Python environment in .venv. Did you forget to activate it?
```

**Comment by angelotrivelli** [comment]
Is there a way to just "force" quarto to use the python virtual environment of the directory in which it has been invoked?
(I am talking here about windows 10/11),

**Reply by cscheid** [reply]
@cderv do you know what could be causing this? User's reporting that activating a virtual environment doesn't change the resolved python binary away from their `.pyenv/...` directory.

**Reply by mcanouil** [reply]
Because you are using VSCode and since it has the ability to activate virtual environment when opening new terminals, could you check if that is working?
Alternatively/concurrently, could you try to reset completely your virtual environment and follow the guide to setup your virtual environment in `env` (not `.venv`[^1]), see <https://quarto.org/docs/projects/virtual-environments.html#using-venv>.

For more about Python virtual environment and VSCode, see <https://code.visualstudio.com/docs/python/environments>.

[^1]: dot files/directories are considered "hidden" files/directories, maybe Quarto is not able to see/use that because of this.

**Reply by angelotrivelli** [reply]
Thanks!
I exited from my ".venv" environment, deleted it, quit vscode, and re-launched vscode, starting fresh with a new terminal. 

Initially, I just have the `hello.qmd` and NOTHING else in project directory. I then create a venv named `env` as instructed and activate it (see screenshot below, looks normal, right?)

<img width="537" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/588784/e97c9e9e-f09d-4870-8b14-c4c01eeb660a"> 

I then issue a `python -m pip install jupyter matplotlib pandas` and it installs dependencies.

When I press the `render` button in vscode it opens a **new** terminal (named `Quarto Preview`), activates my venv, and then issues the quarto preview command...
```
PS C:\projects\quarto_python_vscode> & c:/projects/quarto_python_vscode/env/Scripts/Activate.ps1
(env) PS C:\projects\quarto_python_vscode> quarto preview c:/projects/quarto_python_vscode/hello.qmd --no-browser --no-watch-inputs
```

But I still get exactly the same result as above. It's clearly using my pyenv python path and not the one in `./env` even though it has obviously activated that virtual environment. 

One thing that I don't quite understand is that there's an ampersand, "&", in front of the activation. [Looking it up it](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_operators?view=powershell-7.3&viewFallbackFrom=powershell-7.1#call-operator-) says this is the "call operator" and it means that the command is executed in a "child scope". Honestly, I don't understand what that means but I don't think it's relevant here (I get the same results if I copy-paste the command without the ampersand).

I should add that _even if I exit vscode entirely_, and just do the activation and quarto preview commands in a powershell from windows terminal, it gives the same results (quarto uses my pyenv python and ignores the virtual env).

**Reply by cderv** [reply, marked as answer]
First let me give a workaround that could unblock you: You can set the env var `QUARTO_PYTHON` to a specific python version to force Quarto to use that one. It you set the one from your virtual env, I think it should unblock you. 

That is a workaround I used in the past in the context of this issue with pyenv https://github.com/quarto-dev/quarto-cli/issues/1463
It seems different for you as it seems `quarto check` now find `pyenv`, but I bet there is an issue in the detection process for virtual environmnent on Windows. 

So seems like an issue to me and I need to look into it.

**Reply by angelotrivelli** [reply]
Thanks @cderv! 

I can now make it work by setting the env var, QUARTO_PYTHON, like this: 

```
   $env:QUARTO_PYTHON=(Get-Command python).Path
```

If I press the "render" button again, it gives me a new terminal and I have to re-set the environment var. I will see what I can do to automate setting the environment variable but this is definitely a workaround !!!

**Reply by cderv** [reply]
You need to set it in your global environment. For example, if you are using Powershell as I can see, you can add the line into the profile which is run at start of each terminal

You can find the profile file path in the `$profile` variable of Powershell. Example for me 
```
> echo $profile
C:\Users\chris\Documents\PowerShell\Microsoft.PowerShell_profile.ps1
```

**Reply by angelotrivelli** [reply]
Thanks, yes, that will certainly work. I was looking into perhaps making that environment variable have a "project scope" by somehow hooking it into the activate script. But now that you mention it, having global scope for this quarto-specific env variable is just fine.

**Comment by jgomezdans** [comment]
I have the same issue on Linux and VSCode. VSCode terminal (bash) has variable `QUARTO_PYTHON` set to the system Python, and not to the conda environment where quarto is installed. In the standard system terminal (bash too), `QUARTO_PYTHON` isn't defined, and it works fine. Anyone can figure out how to fix this?

**Reply by cderv** [reply]
> I have the same issue on Linux and VSCode. 

By same issue, do you mean it also happens with `pyenv` with virtual env activated ? 

> VSCode terminal (bash) has variable QUARTO_PYTHON set to the system Python, and not to the conda environment where quarto is installed.

VSCODE Quarto Extension for Python document will leverage the Python Extension and the selected interpreter. It will use that one. 
Did you check the right one is selected in your extension ? You can change the Python interpreter selected in VSCODE through the python extension configuration.

**Reply by jgomezdans** [reply]
Thanks! I changed the Python extension interpreter, and I got quarto working.

**Comment by rnckp** [comment]
I ran into a similar issue with VS Code on MacOS. I use conda for all my environments. Quarto would render with the global interpreter rather than with a specific conda env.

I solved it by these steps:
- Invoke the `Command Palette` in VSC (either from the `View` menu or with `Shift-CMD-P`)
- Select `Python: Select interpreter` 
- Select the conda environment you want to use with Quarto. 
- Render your Quarto document

PS: I also had to install [jupyter_contrib_extensions](https://jupyter-contrib-nbextensions.readthedocs.io/en/latest/install.html) to allow for code execution during render of Quarto docs.

**Reply by moneuron** [reply]
saved me a lot of time. thanks!

**Reply by liganega** [reply]
This is the answer for me.

**Comment by th0ger** [comment]
> I've been using poetry to work with python projects, but I couldn't get it to work with this quarto example. So in order to increase the likelyhood of getting help, I am following the example in getting-started as closely as possible (and temporarily abandoning poetry).

Same issue with Poetry here. I found out that if Poetry creates in-project .venv and this is added to .gitignore, VS Code Python can't find the python .venv!
(I have yet to find if this is a know/solved issue.)
Nothing to do with quarto.

```bash
$ poetry config --list | grep virtualenvs.in-project
virtualenvs.in-project = true

$ poetry init
[tool.poetry]
name = "debug_my_venv"
version = "0.1.0"
description = ""
authors = ["..."]
readme = "README.md"
packages = [{include = "debug_my_venv"}]

[tool.poetry.dependencies]
python = "^3.11"

[build-system]
requires = ["poetry-core"]
build-backend = "poetry.core.masonry.api"

$ poetry shell
$ poetry install
(.venv/ generated)
$ echo ".venv/" > .gitignore
$ touch foo.py
$ code .
```
Lower right corner of VS Code show that *system* python interpreter "3.11.4 64-bit" (`/bin/python`) has been auto-seleted.

"> Python: Select Interpreter" also recommmeds `/bin/python`, and **.venv/ is not in selection list!**

Now, repeat the entire test in a new folder, but *without creating .gitignore*. The interpeter `3.11.4 (.venv: poetry)` is recommended and auto-selected.

```
